charms/tensorboards-web-app/src/charm.py

```diff
@@ -22,119 +36,195 @@ def __init__(self, msg: str, status_type=None):
-                                {
-                                    "apiGroups": ["storage.k8s.io"],
-                                    "resources": ["storageclasses"],
-                                    "verbs": ["get", "list", "watch"],
-                                },
-                                {
-                                    "apiGroups": ["kubeflow.org"],
-                                    "resources": ["poddefaults"],
-                                    "verbs": ["get", "list", "watch"],
-                                },
-                            ],
-                        }
-                    ]
+    @property
+    def _tensorboards_web_app_layer(self) -> Layer:
+        """Create and return Pebble framework layer."""
+        exec_command = (
+            "gunicorn"
+            " -w 3"
+            f" --bind 0.0.0.0:{self._http_port}"
+            " --access-logfile"
+            " - entrypoint:app"
+        )
+
+        layer_config = {
+            "summary": "jupyter-web-app layer",
+            "description": "Pebble config layer for jupyter-web-app",
+            "services": {
+                self._container_name: {
+                    "override": "replace",
+                    "summary": "Entrypoint of jupyter-web-app image",
+                    "command": exec_command,
+                    "startup": "enabled",
+                    "environment": self._env_vars,
+                    "on-check-failure": {"up": "restart"},
+                }
+            },
+            "checks": {
+                "up": {
+                    "override": "replace",
+                    "period": "30s",
+                    "http": {"url": f"http://localhost:{self._http_port}"},
                 },
-                "containers": [
-                    {
-                        "name": "tensorboards-web-app",
-                        "imageDetails": image_details,
-                        "envConfig": {
-                            "USERID_HEADER": "kubeflow-userid",
-                            "USERID_PREFIX": "",
-                            "APP_PREFIX": "/tensorboards",
-                            "APP_SECURE_COOKIES": str(config["secure-cookies"]),
-                            "BACKEND_MODE": config["backend-mode"],
-                        },
-                        "ports": [{"name": "http", "containerPort": config["port"]}],
-                    }
-                ],
             },
-        )
+        }
+
+        return Layer(layer_config)
 
-        self.model.unit.status = ActiveStatus()
+    def _on_install(self, _) -> None:
+        """Perform installation only actions."""
+        try:
+            # Deploy K8S resources to speed up deployment
+            self._deploy_k8s_resources()
+        except CheckFailed as err:
+            self._log_and_set_status(err.status)
+        return
 
-    def _configure_mesh(self, interfaces):
+    def _on_remove(self, _) -> None:
```

Comment:
Totally agree here.

We've already merged this in the Tensorboard Controller: https://github.com/canonical/kubeflow-tensorboards-operator/blob/fba41aae289e978c99be6a1f02cb01958f66a6e7/charms/tensorboard-controller/src/charm.py#L211-L222

Note that you won't have to collect all errors like we do in the controller (due to the fact that we have 2 resource handlers there), since this functionality is now covered in `chisme`:
* https://github.com/canonical/charmed-kubeflow-chisme/pull/62

Reply:
I think this still applies, I'm +1 on this suggestion.

Reply:
@DnPlas It does, I hadn't used the Files changed view for viewing comments and it turns out that I had missed some by Andrew which didn't appear in the Conversation tab.

Reply:
Thanks for the suggestion, I 'm pushing a commit for this.